mission/mission_planner.cpp

```diff
@@ -0,0 +1,1038 @@
+        {
+            inserted_waypoint_ = Waypoint(  MAV_FRAME_LOCAL_NED,
+                                            MAV_CMD_NAV_TAKEOFF,
+                                            1,
+                                            0.0f,
+                                            0.0f,
+                                            0.0f,
+                                            coord_conventions_get_yaw(ahrs_.qe),
+                                            ins_.position_lf()[X],
+                                            ins_.position_lf()[Y],
+                                            navigation_.takeoff_altitude);
+            insert_ad_hoc_waypoint(inserted_waypoint_);
+            require_takeoff_ = false;
+        }
+
+        if (current_mission_handler_ != NULL)
+        {
+            // Handle current mission
+            bool ret = current_mission_handler_->update(*this);
+
+            // Check if we should be switch states
+            // This is for automatic advancement (states can also advance from callback
+            // functions from GCS or remote)
+            if (ret == 1)
+            {
+                switch (internal_state())
+                {
+                case STANDBY:
+                    // DONT ADVANCE IF FINISHED ON GROUND
+                    break;
+
+                case PREMISSION: // After takeoff, continue with next mission item
+                    {
+                        // Advance mission item
+                        navigation_.set_start_wpt_time();
+                        navigation_.set_waiting_at_waypoint(false);
+                        waypoint_handler_.advance_to_next_waypoint();
+
+                        // Set mission handler
+                        switch_mission_handler(waypoint_handler_.current_waypoint());
+
+                        print_util_dbg_print("Autocontinue from takeoff towards waypoint Nr");
+                        print_util_dbg_print_num(waypoint_handler_.current_waypoint_index(),10);
+                        print_util_dbg_print("\r\n");
+
+                        // Send message
+                        mavlink_message_t msg;
+                        mavlink_msg_mission_current_pack(mavlink_stream_.sysid(),
+                                                         mavlink_stream_.compid(),
+                                                         &msg,
+                                                         waypoint_handler_.current_waypoint_index());
+                        mavlink_stream_.send(&msg);
+                        break;
+                    }
+
+                case MISSION: // After mission item, continue with next mission item
+                    {
+                        // Advance to next waypoint
+                        navigation_.set_start_wpt_time();
+                        navigation_.set_waiting_at_waypoint(false);
+                        waypoint_handler_.advance_to_next_waypoint();
```

Comment:
I left the autocontinue check with the handlers in case someone created a handler that wanted to ignore the autocontinue for some specific reason (e.g. have a timeout or something), but we could easily make the autocontinue to be enforced by adding a check in mission_planner.